src/index.ts

```diff
@@ -0,0 +1,384 @@
+        this.emit('failed', error);
+      }
+      this.serial_error.bind(this);
+    });
+    this.serial.on('close', () => {
+      this.options.debug && console.debug('ryderserial close');
+      this.emit('close');
+      clearInterval(this[reconnect_symbol]);
+      if (!this.closing)
+        this[reconnect_symbol] = setInterval(this.open.bind(this), this.options.reconnectTime);
+    });
+    this.serial.on('open', () => {
+      this.options.debug && console.debug('ryderserial open');
+      // reset or keep?
+      // this[train_symbol] = [];
+      // this[state_symbol] = STATE_IDLE;
+      // this[lock_symbol] = [];
+      clearInterval(this[reconnect_symbol]);
+      this.emit('open');
+      this.next();
+    });
+  };
+
+  close(): void {
+    if (this.closing)
+      return;
+    this.closing = true;
+    this.clear();
+    this.serial?.close();
+    clearInterval(this[reconnect_symbol]);
+    this.serial = undefined;
+  };
+
+  locked(): boolean {
+    return !!this[lock_symbol].length;
+  };
+
+  lock(): Promise<void> {
+    this.options.debug && console.debug('ryderserial lock');
+    if (!this[lock_symbol].length) {
+      // TODO: fix this[lock_symbol]... why is it holding 'false'?
+      // the line below used to read `this[lock_symbol].push(false);` confirm that the line below is safe?
+      this[lock_symbol].push(Promise.resolve);
+      return Promise.resolve();
+    }
+    return new Promise((resolve, reject) => {
+      this[lock_symbol].push(resolve)
+    });
+  };
+
+  unlock(): void {
+    if (this[lock_symbol].length) {
+      this.options.debug && console.debug('ryderserial unlock');
+      var resolve = this[lock_symbol].shift();
+      if (!resolve && this[lock_symbol].length)
+        resolve = this[lock_symbol].shift();
+      resolve && resolve();
+    }
+  };
+
+  sequence(callback: (value: any) => Promise<any>) {
+    if (typeof callback !== 'function' || callback.constructor.name !== 'AsyncFunction')
+      return Promise.reject(new Error('ERROR_SEQUENCE_NOT_ASYNC'));
+    return this.lock().then(callback).finally(this.unlock.bind(this));
+  };
+
+  send(data: string | number, prepend?: boolean): Promise<void> {
+    if (!this.serial || !this.serial.isOpen)
+      return Promise.reject(new Error('ERROR_DISCONNECTED'));
+    if (typeof data === 'number')
+      data = String.fromCharCode(data);
+    this.options.debug && console.debug('queue data for Ryder: ' + data.length + ' byte(s)', Buffer.from(data).toString("hex"));
+    return new Promise((resolve, reject) => {
+      const c: TrainEntry = [data as string, resolve, reject, false, ''];
+      prepend ? this[train_symbol].unshift(c) : this[train_symbol].push(c);
+      this.next();
+    });
+  };
+
+  next(): void {
+    if (this[state_symbol] === STATE_IDLE) {
+      if (!this[train_symbol].length)
+        return;
+      if (!this.serial || !this.serial.isOpen) {
+        const [, , reject] = this[train_symbol][0];
+        this.clear();
```

Comment:
I wonder if we can to keep the queued commands for when the device reconnects or not. I think this is fine for now and we can decide based on future use-cases. I would imagine that an app dev would show a warning if the device becomes disconnected and for the user to restart whatever process it was in.

Reply:
Oh, this is an interesting thought...

Agreed that this can wait for later. Especially because I think the way we treat `this[train_symbol]` currently is potentially the weakest part of this code right now... I'm still motivated to refactor it into its own class but think that should wait for a different PR (this one is gettin' big).